packages/hardhat/test/e2e/crab-Migration.ts

```diff
@@ -273,4 +322,702 @@ describe("Crab Migration", function () {
+
+
+                expect(wdiv(collatV1, shortV1).gt(wdiv(collatV2, shortV2))).to.be.false
+                expect(isFlashMigrate).to.be.false
+                expect(ethToGetFromV1.gt(ethNeededForV2)).to.be.false
+
+                const oSqthPrice = await oracle.getTwap(wethOsqthPoolAddress, squeethAddress, wethAddress, 1, false)
+
+                const numerator = ethToGetFromV1.sub(wmul(v1oSqthToPay, oSqthPrice).mul(101).div(100)) // 1% slippage
+                const denominator = one.sub(wdiv(oSqthPrice.mul(101).div(100), wdiv(collatV2, shortV2)))
+                const ethToFlashLoan = wdiv(numerator, denominator)
+                const expectedSqthFromV2 = wdiv(wmul(ethToFlashLoan, shortV2), collatV2)
+                const v1SqthToFlashWithdraw = v1oSqthToPay.sub(expectedSqthFromV2)
+                const maxEthToPay = wmul(v1SqthToFlashWithdraw, oSqthPrice).mul(1005).div(1000) // .1% slippage
+                const ethToGetFromWithdraw = wdiv(wmul(expectedSqthFromV2, collatV1), shortV1)
+                const ethToGetFromFlashWithdraw = wdiv(wmul(v1SqthToFlashWithdraw, collatV1), shortV1).sub(maxEthToPay)
+                const expectedEth = ethToGetFromFlashWithdraw.add(ethToGetFromWithdraw).sub(ethToFlashLoan)
+
+                const depositShare = wdiv(ethToFlashLoan, collatV2.add(ethToFlashLoan))
+                const expectedV2Shares = wdiv(wmul(depositShare, crabV2Supply), one.sub(depositShare))
+
+                const tx1 = await crabStrategyV1.connect(d3).approve(crabMigration.address, ethers.constants.MaxUint256);
+                gasPaid = await getGasPaid(tx1)
+                // Don't flash deposit and return back ETH. For the sake of simplicity. Flash deposit is already tested in previous tests
+                const tx2 = await crabMigration.connect(d3).flashMigrateAndWithdrawFromV1toV2(crabV1SharesBefore, 0, ethToFlashLoan, maxEthToPay)
+                gasPaid = gasPaid.add(await getGasPaid(tx2))
+
+                const crabV1SharesAfter = await crabStrategyV1.balanceOf(d3.address)
+                const crabV2SharesAfter = await crabStrategyV2.balanceOf(d3.address)
+                const crabV1SharesInMigration = await crabStrategyV1.balanceOf(crabMigration.address)
+                const crabV2SharesInMigration = await crabStrategyV2.balanceOf(crabMigration.address)
+                const userEthBalanceAfter = await provider.getBalance(d3.address)
+                const squeethBalance = await oSqth.balanceOf(crabMigration.address)
+
+
+                expect(crabV1SharesAfter).to.be.equal('0')
+                expect(crabV1SharesInMigration).to.be.equal('0')
+                expect(crabV2SharesInMigration).to.be.equal('0')
+                expect(crabV2SharesAfter).to.be.equal(expectedV2Shares)
+                // Can't predict exact because of slippage so should be greater or equal to expected balance
+                expect(userEthBalanceAfter.gte(userEthBalance.add(expectedEth).sub(gasPaid))).to.be.true
```

Comment:
same comment as above (think only need to probably test 1x, but think we should get 1 exact result)